python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -652,3 +658,220 @@ def _isin_datetimelike(
 
     res = lhs._obtain_isin_result(rhs)
     return res
+
+
+def date_range(
+    start=None,
+    end=None,
+    periods=None,
+    freq=None,
+    tz=None,
+    normalize=False,
+    name=None,
+    closed=None,
+):
+    """Return a fixed frequency DatetimeIndex.
+
+    Returns the range of equally spaced time points (where the difference
+    between any two adjacent points is specified by the given frequency)
+    such that they all satisfy `start` <[=] x <[=] `end`, where the first one
+    and the last one are, resp., the first and last time points in that range
+    that are valid for `freq`.
+
+    Parameters
+    ----------
+    start : str or datetime-like, optional
+        Left bound for generating dates.
+
+    end : str or datetime-like, optional
+        Right bound for generating dates.
+
+    periods : int, optional
+        Number of periods to generate.
+
+    freq : DateOffset
```

Comment:
Some low hanging fruit here might be to support some string values through `cudf.DateOffset._from_freqstr`. This is often the API I used to reach for as a user since it was the fastest to type:

```
data = pd.date_range('2011-01-01', '2011-01-10', freq='1D')
```

Reply:
I think we can support `D` relatively easily. Beyond that the frequency string often get defaulted to some anchored offset or normalized as begin/end of month/year etc.
Example:
`W` defaults as anchored offset:
```python
pd.date_range('2011-01-01 05:43:10', '2011-01-10 16:43:10', freq='1W')
DatetimeIndex(['2011-01-02 05:43:10', '2011-01-09 05:43:10'], dtype='datetime64[ns]', freq='W-SUN')
```
`M` means month end:
```python
pd.date_range('2011-01-01 05:43:10', '2011-03-10 16:43:10', freq='M')
DatetimeIndex(['2011-01-31 05:43:10', '2011-02-28 05:43:10'], dtype='datetime64[ns]', freq='M')
```

Reply:
Note that there's also a slight mismatch between [offset alias](https://pandas.pydata.org/docs/user_guide/timeseries.html#offset-aliases) and codes in cudf.DateOffset. E.g. `m` is mapped to minutes in cudf.DateOffset, but month in offset alias. `T`/`min` is mapped to minutes but not supported in cudf.DateOffset. Are we allowed to make these deviation here?

Reply:
[Update] Ok. So I made a mapping from the offset aliases to the codes in `cudf.DateOffset`. We still can only support a limit range of what's available in pandas due to reasons above. And we can't support `L` and `ms` due to https://github.com/pandas-dev/pandas/issues/43371